ir/src/integrity_stmts/graph.rs

```diff
@@ -90,6 +90,7 @@ impl AlgebraicGraph {
         &mut self,
         symbol_table: &SymbolTable,
         expr: IntegrityExpr,
+        row_offsets: &mut Vec<usize>,
```

Comment:
As noted, I think we should change the output signature rather than the input. I also think we should  just track the highest row offset, since that defines the constraint. I don't see why we care about all of the row offsets at this point, and it will probably be simpler not to keep them all unless we actually have to. We should be able to just treat the row_offset the same way that we deal with the `trace_segment`, i.e. pass it up from the leaves and then take the max each time we do an operation with 2 sub-expressions such as `Add`

Let me know what you think and if there's something I overlooked.

Also, since we'd now be returning a tuple of 3 elements, it probably makes sense to have a convenience type alias for `(TraceSegment, NodeIndex, usize)`. It might also make sense to have an alias `DEFAULT` equal to `CURRENT_ROW`, so that we're not talking about rows when dealing with variables. That's just a stylistic thought though.

Reply:
That makes sense. I was going by the signature @bobbinth posted in the issue:
```
struct ConstraintRoot {
    index: NodeIndex,
    offsets: Vec<usize>,
}
```
I thought maybe Bobbin suggested that as keeping track of all the row offsets had utility for other projects that jump more than one rows but maybe I misunderstood something?

Reply:
I think tracking all offsets could be useful in the future (when we allow constraints which work with more than 2 rows) because the constraints which access the same set of rows may be treated differently (though, I can't come up with concrete examples now).

In general, however, this structure will probably change once boundary constraints are integrated into the graph. Specifically, for each constraint we'll need to define a set of rows it works with and just offsets would not be enough (i.e., for boundary constraints, instead of offsets we'd need to specify whether it access the first or the last row). In Winterfell, there is a [ConstraintDivisor](https://github.com/facebook/winterfell/blob/main/air/src/air/divisor.rs#L13) struct which does something similar - but I think in our case we can probably use something simpler.

Reply:
We could have something like `ConstraintDomain` enum:
```Rust
pub enum ConstraintDomain {
  FirstRow,           // for boundary constraints against the first row
  LastRow,            // for boundary constraints against the last row
  EveryRow,           // for validity constraints
  EveryWindow(usize), // for transition constraints
}
```
For `EveryWindow` the internal value specifies the size of the evaluation frame. E.g., for transition constraints over two consecutive rows the internal value would be `2`. I used the term `window` as this is what Rust uses for [similar concepts](https://doc.rust-lang.org/std/primitive.slice.html#method.windows) - but I'm not too happy with it. So, open to other suggestions.

If the need arises, we could add other variants to this enum.

Reply:
hm follow-up, I guess you're talking about a case where we access e.g. 3 rows with offsets 1, 3, 5. So in this case we'd want to spot matching `ConstraintRoots` that also access the rows with these offsets, and that's why we'd want all of those offsets?

But then our window would be from row 0 to row 5, with `ConstraintDomain::EveryWindow(6)`, but that `ConstraintDomain` doesn't have the the information about the particular set either. Or are you thinking about capturing it by having multiple `ConstraintDomain` associated with the `ConstraintRoot`, e.g. `ConstraintDomain::EveryWindow(1)`, `ConstraintDomain::EveryWindow(3)`, `ConstraintDomain::EveryWindow(5)`

Reply:
I guess it feels like unnecessary complexity for something for which we don't have a concrete use-case, and it wouldn't be hard to change this in the future if/when we do have a concrete use, at which point we'd have more information about what would work best.

Reply:
Yes, the intent was to differentiate between constraints which require frame size 3, but one of them access all 3 rows, while the other one access only rows 0 and 2. But as you said, we don't need this for now - and probably makes sense to add this when/if we need it.

I like `Frame`! So, the enum would be:
```Rust
pub enum ConstraintDomain {
  FirstRow,           // for boundary constraints against the first row
  LastRow,            // for boundary constraints against the last row
  EveryRow,           // for validity constraints
  EveryFrame(usize),  // for transition constraints; internal value is size of the frame
}
```
We don't need to use this now - but would make sense to do when we merge boundary and transition constraints in the same graph. Then, `ConstraintRoot` struct would look something like:
```Rust
struct ConstraintRoot {
    index: NodeIndex,
    domain: ConstraintDomain,
}
```

Reply:
We'll add this enum in #106